File: leetcode-solutions/0543-diameter-of-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxdepth(TreeNode* root,int& totmax){
        if(root==nullptr)return 0;
        int lh=maxdepth(root->left,totmax);
        int rh=maxdepth(root->right,totmax);
        int curdia=lh+rh;
        totmax=max(totmax,curdia);
        return 1+max(lh,rh);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        if(root==nullptr)return 0;
        int totmax=0;
        maxdepth(root,totmax);
        return totmax;
    }
};

File: leetcode-solutions/0543-diameter-of-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxH=0;
    public int maxHeight(TreeNode root){
        if(root==null)return 0;
        int left=maxHeight(root.left);
        int right=maxHeight(root.right);
        maxH=Math.max(left+right,maxH);
        return 1+Math.max(left,right);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        maxHeight(root);
        return maxH;
    }
}
